### PR TITLE
Added documentation for nearest-neighbor-gradient [cont]

### PR DESCRIPTION
**Pull request by davidscn**
Follow-up to #166

### CODE REVIEW

**Comment by boris-martin** [comment]
"There are three main variants of mapping" should become four I think.

File: pages/docs/couple-your-code/couple-your-code-gradient-data.md [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+  ... ,
+  vnx_dx, vnx_dy, vny_dx, vny_dy  )  
+```
+
+When using the `rowsFirst` parameter, the following format is required:
+
+```cpp
+( v0x_dx, v0y_dx, v1x_dx, v1y_dx, ... , vnx_dx, vny_dx,
+  v0x_dy, v0y_dy, v1x_dy, v1y_dy, ... , vnx_dy, vny_dy  )
+```
```
Comment (davidscn) [inline pages/docs/couple-your-code/couple-your-code-gradient-data.md]: This goes beyond the documentation part, but considering that the 'gradient matrix' is a Jacobian, I think the `rowFirst` argument should be a column first, as the Jacobian contains the derivative in a particular direction in a column, right?

File: pages/docs/couple-your-code/couple-your-code-gradient-data.md [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    int             valueIndex,
+    const double*   gradientValues,
+    bool            rowsFirst = false );
+
+void writeBlockVectorGradientData (
+      int           dataID,
+      int           size,
+      const int*    valueIndices,
+      const double* gradientValues,
+      bool          rowsFirst = false );
```
Comment (uekerman) [inline pages/docs/couple-your-code/couple-your-code-gradient-data.md]:
```suggestion
void writeScalarGradientData (
    int             dataID,
    int             valueIndex,
    const double*   gradientValues  );

void writeBlockScalarGradientData (
    int           dataID,
    int           size,
    const int*    valueIndices,
    const double* gradientValues  );

void writeVectorGradientData (
    int             dataID,
    int             valueIndex,
    const double*   gradientValues,
    bool            rowsFirst = false );

void writeBlockVectorGradientData (
    int           dataID,
    int           size,
    const int*    valueIndices,
    const double* gradientValues,
    bool          rowsFirst = false );
```

File: pages/docs/couple-your-code/couple-your-code-gradient-data.md (hunk 1 above)
Comment (uekerman) [reply pages/docs/couple-your-code/couple-your-code-gradient-data.md]: > Jacobian contains the derivative in a particular direction in a column, right?

image

yes.
But here, we have a 3D tensor, right? So "rowFirst" or "columnFirst" will never tell the full story, right?

File: pages/docs/couple-your-code/couple-your-code-gradient-data.md [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+  // write gradient data
+  if (isGradientDataRequired(dataID)){
+    computeStressGradient(stressGradient)
+    precice.writeBlockVectorGradientData(stressID, vertexSize, vertexIDs, stressGradient);
+  }
+
+  precice_dt = precice.advance(dt);
+}
+[...]
+```
```
Comment (uekerman) [inline pages/docs/couple-your-code/couple-your-code-gradient-data.md]: Let's shorten this some more:

```suggestion
```cpp
precice::SolverInterface precice("FluidSolver","precice-config.xml",rank,size); // constructor

int dim = precice.getDimensions();
int meshID = precice.getMeshID("FluidMesh");
[...] 
precice.setMeshVertices(meshID, vertexSize, coords, vertexIDs);

int stressID = precice.getDataID("Stress", meshID);
double* stress = new double[vertexSize * dim];

// create gradient data
double* stressGradient = new double[vertexSize * dim * dim]
[...]
precice_dt = precice.initialize();

while (not simulationDone()){ // time loop
  precice.readBlockVectorData(displID, vertexSize, vertexIDs, displacements);
  setDisplacements(displacements);
  [...]
  solveTimeStep(dt);
  computeStress(stress);

  precice.writeBlockVectorData(stressID, vertexSize, vertexIDs, stress);

  // write gradient data
  if (isGradientDataRequired(dataID)){
    computeStressGradient(stressGradient)
    precice.writeBlockVectorGradientData(stressID, vertexSize, vertexIDs, stressGradient);
  }

  precice_dt = precice.advance(dt);
}
[...]
```
```

File: pages/docs/couple-your-code/couple-your-code-gradient-data.md [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+
+```xml
+<?xml version="1.0"?>
+
+<precice-configuration>
+
+  <solver-interface dimensions="3" experimental="on">
+
+    <data:vector name="Stress" gradient="on"/>
+    <data:vector name="Displacements" gradient="on" />
```
Comment (uekerman) [inline pages/docs/couple-your-code/couple-your-code-gradient-data.md]:
```suggestion
    <data:vector name="Displacements" />
```

Above we only use gradients for displacements. I think such a mixed example could be good anyway.

File: pages/docs/couple-your-code/couple-your-code-gradient-data.md [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+
+    <participant name="FluidSolver">
+      <use-mesh name="FluidMesh" provide="yes"/>
+      <use-mesh name="StructureMesh" from="SolidSolver"/>
+      <write-data name="Stress" mesh="FluidMesh"/>
+      <read-data  name="Displacements" mesh="FluidMesh"/>
+      <mapping:nearest-neighbor-gradient direction="write" from="FluidMesh"
+                                to="StructureMesh" constraint="consistent"/>
+      <mapping:nearest-neighbor-gradient direction="read" from="StructureMesh"
+                                to="FluidMesh" constraint="consistent"/>
```
Comment (uekerman) [inline pages/docs/couple-your-code/couple-your-code-gradient-data.md]:
```suggestion
      <mapping:nearest-neighbor direction="read" from="StructureMesh"
                                to="FluidMesh" constraint="consistent"/>
```

**Review by uekerman — APPROVED** [review]
Good to go except minor tweaks.

> "There are three main variants of mapping" should become four I think.

yes :+1:

**Comment by davidscn** [comment]
This is now ready to merge once https://github.com/precice/precice.github.io/pull/169 will be merged.

**Review by uekerman — APPROVED** [review]

**Pull request merged**